composer/trainer/trainer.py

```diff
@@ -1518,14 +1518,16 @@ def _train_loop(self) -> None:
                 if use_grad_scaling:
                     self.state.scaler.update()
 
+                # total_loss can be None if gradient scaling failed
                 if total_loss is not None:
-                    if not isinstance(total_loss, torch.Tensor):
-                        total_loss = self._device.tensor_to_device(torch.tensor([total_loss]))
-
-                    # total_loss can be None if gradient scaling failed
-                    dist.all_reduce(total_loss, reduce_operation='SUM')
-                    full_loss = total_loss.cpu().item()
-                    self.logger.data_batch({'loss/train': full_loss / dist.get_world_size()})
+                    if isinstance(total_loss, dict):
+                        map_collection(total_loss, dist.all_reduce)
+                        total_loss = {k: loss.cpu().item() / dist.get_world_size() for k, loss in total_loss.items()}
+                        self.logger.data_batch(total_loss)
+                    if isinstance(total_loss, torch.Tensor):
+                        dist.all_reduce(total_loss, reduce_operation='SUM')
+                        full_loss = total_loss.cpu().item()
+                        self.logger.data_batch({'loss/train/total': full_loss / dist.get_world_size()})
```

Comment:
I thought total_loss should always be a dict, but pyright wasn't convinced. I added this check to appease pyright, does this look fine?